### PR TITLE
ThreadedActionListener always wraps listener to preserve context

### PR DESCRIPTION
**Pull request by jaymode**
The ThreadedActionListener may not be invoked by the same thread that creates it. When this happens, the ThreadContext from the thread that created the ThreadedActionListener should be used. This change makes the ThreadedActionListener always wrap the ActionListener in a ContextPreservingActionListener to prevent accidental loss of the ThreadContext.

Relates #23349

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/support/ThreadedActionListener.java [hunk 1]
```diff
@@ -81,7 +81,7 @@ public ThreadedActionListener(Logger logger, ThreadPool threadPool, String execu
         this.logger = logger;
         this.threadPool = threadPool;
         this.executor = executor;
-        this.listener = listener;
+        this.listener = ContextPreservingActionListener.wrap(listener, threadPool.getThreadContext(), false);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/ThreadedActionListener.java]: I know we talked about it but reflecting on this, I think it's too implicit to wrap things here. Better be explicit and put this as a parameter type and let the caller be aware and decide.

**Review by bleskes — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/elasticsearch/action/support/AbstractListenableActionFuture.java [hunk 2]
```diff
@@ -50,7 +50,8 @@ public void addListener(final ActionListener<T> listener) {
     }
 
     public void internalAddListener(ActionListener<T> listener) {
-        listener = new ThreadedActionListener<>(logger, threadPool, ThreadPool.Names.LISTENER, listener, false);
+        listener = new ThreadedActionListener<>(logger, threadPool, ThreadPool.Names.LISTENER,
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/AbstractListenableActionFuture.java]: hmm.. is this correct? It's debatable what we should do here. Should we restore the context in which the listener was added or should we use the context of the thread that made the request and also keep the response headers. The way this is set up now, consumers don't have a choice. I'm starting to doubt if we should do this... @s1monw wdyt?

Comment (jaymode) [inline core/src/main/java/org/elasticsearch/action/support/AbstractListenableActionFuture.java]: I agree and am beginning to think we should skip this change altogether for now

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/AbstractListenableActionFuture.java]: > think we should skip this change altogether

yeah. Let's just drop this. Thanks for trying..

**Pull request closed without merging**